### ISSUE TITLE
Check against custom property registration

### ISSUE THREAD

**Issue by shgysk8zer0** [stylelint/stylelint#5061]
> Check custom property types against `@property` rules or custom property registration.

> What solution would you like to see?

Add a rule allowing to specify custom property registrations such that `--primary-color` must be a `"<color>"`. This rule should accept a file path accepting a CSS file containing [`@property`](https://developer.mozilla.org/en-US/docs/Web/CSS/@property) declarations or a JSON file that is compatible with [`CSS.regsiterProperty()`](https://developer.mozilla.org/en-US/docs/Web/API/CSS/RegisterProperty). It may further check that all usage of registered custom properties are used correctly, such that `height: var(--primary-color)` is invalid.

### CSS
```css
@property --primary-color {
  syntax: "<color>";
  inherits: true;
  initial-value: red;
}

/* ...*/
```

### JSON
```json
[{
  "name": "--primary-color",
  "syntax": "<color>",
  "inherits": true,
  "initial-value": "red"
}]
```

**Comment by jeddy3** [comment]
@shgysk8zer0 Thanks for the request and for using the template.

It's a very interesting proposal. 

[CSSTree](https://github.com/csstree/csstree/blob/master/README.md) has a lexer that can validate the syntax of declaration values, e.g. [for the `<color>` type](https://csstree.github.io/docs/syntax/#Type:color). The existing [stylelint-csstree-validator plugin](https://github.com/csstree/stylelint-validator) has [options](https://github.com/csstree/stylelint-validator#options) that get pretty close to the proposal, i.e. specifying syntax for properties unknown to the CSS specifications.

> This rule should accept a file path accepting a CSS file containing [`@property`](https://developer.mozilla.org/en-US/docs/Web/CSS/@property) declarations or a JSON file that is compatible with [`CSS.regsiterProperty()`](https://developer.mozilla.org/en-US/docs/Web/API/CSS/RegisterProperty). 

The [stylelint-value-no-unknown-custom-properties plugin](https://github.com/csstools/stylelint-value-no-unknown-custom-properties) has prior art for this in its [`importFrom` option](https://github.com/csstools/stylelint-value-no-unknown-custom-properties#importfrom).

I suggest reaching out to the maintainers of the [stylelint-csstree-validator plugin](https://github.com/csstree/stylelint-validator) to see if they're happy to expand their plugin to read syntax information for custom properties from `@property` at-rules and `PropertyDefinition` dictionary objects. 

If they're not, then I suggest [creating a plugin](https://stylelint.io/developer-guide/plugins) to flesh out the implementation. It can be modelled on the [stylelint-csstree-validator plugin](https://github.com/csstree/stylelint-validator), in that it uses PostCSS to parse the larger constructs and CSSTree to parse the declaration value (using its [`context` option](https://github.com/csstree/csstree/blob/master/docs/parsing.md#context)).  

I'll label as discussion in case anyone else has thoughts on this.

**Comment by evilebottnawi** [comment]
@jeddy3 sometimes I think we need migrate on new parser

**Comment by jeddy3** [comment]
I know what you're saying, @alexander-akait. I think that too sometimes.

The granularity of the CSSTree AST blows my mind. For example, [this detailed AST](https://astexplorer.net/#/gist/e83a13903adbd399d068ad84c4997cf1/a97571b65a4f75a750441eb5ed29ef0dea1be557) has:

- declarations within at-supports
- media query lists
- an+b micro syntax within selectors

Its syntax validation boggles the mind too.

However, PostCSS is also amazing. Its coarseness is one of its strengths as it's forgiving of non-standard syntax. I don't think we'd be able to support the styling languages we do without it.

I'm not sure where that leaves us. If we were building stylelint today, I suspect we'd pretty print the AST (rather than fixing stylistic things piecemeal) and we'd use the most detailed CSS parser we can find, and then cross the non-standard syntax bridge when we come to it. Right now, though, I think using PostCSS means the most number of users, regardless of which styling language they prefer, can lint their code.

**Comment by brookjordan** [comment]
I would love to bump this up. As Typescript rises in popularity, new developers are expecting the same kind of help with CSS. Getting these kinds of typings in eslint would be a massive help, and a great step forward.
